Enable no-checkpoint Tapenade adjoint generation with STREAMICE for fast adjoint computation

dngoldberg opened the pull request:
## What changes does this PR introduce?
This PR introduces a facility to pass command line flags to Tapenade (including "-defaultnocheckpoint") and additionally provides FWD and BWD templates of manually generated derivative code.

## What is the current behaviour? 
By default, when TAPENADE is invoked to create the adjoint, the behaviour is to produce adjoint subroutines (*_B) which first call the primal (original code) then adjoint of each subroutine called within the original subroutine. This optimises memory usage but greatly increases computation time relative to the undifferentiated code. In other words, checkpointing is "turned on" everywhere.

Turning off checkpointing increases memory but reduces runtime. It is not possible to do this without either (a) invasive per-subroutine pragmas or (b) a command-line flag (-defaultnocheckpoint) passed to the tapenade command. Note that doing so also initiates the generation of _FWD and _BWD versions of every subroutine. For subroutines that are manually generated, such subroutines must be written. They are related to _B subroutines, but include code for managing the adjoint tape memory.

## What is the new behaviour 

This pull request makes a modification to genmake2 which allows a flag "taparg" and an argument allowing a flag, namely `defaultnocheckpoint` to be passed to Tapenade. For example, a genmake2 call might look like:

`genmake2 -ds -m make -tap -mods=../code_tap -taparg="-defaultnocheckpoint"`

Such a flag is very beneficial to STREAMICE and gives great runtime advantage without unsustainable growth in memory needs -- see Fig 4 of https://arxiv.org/html/2405.15590v1#S5. (A related pull request [927](https://github.com/MITgcm/MITgcm/pull/927), which implements an algorithm that lessens the memory requirements of the adjoint model, will help with this.) For instance, the `halfpipe_streamice` verification, which takes has 3 time steps, evaluates each time step 3 times without the `defaultnocheckpont` flag, but only once with it set.

To facilitate `defaultnocheckpoint`, the _FWD and _BWD versions of the following subroutines are written:
```
GLOBAL_SUM_TILE_RL
EXCH1_RL
EXCH1_RL_CUBE
EXCH1_RS
ACTIVE_READ_XY
ACTIVE_WRITE_XY
```
without which compliing with the Makefile generated by the above genmake2 would lead to undefined references.

## Does this PR introduce a breaking change? 
Yes, On github, the halfpipe_streamice test now passes with only 10 digits (likely due to a difference in order of operations), and the results file may need reproducing

## Other information:
- pull request [927](https://github.com/MITgcm/MITgcm/pull/927), mentioned above, is related
- this PR addresses point 15 of issue [735](https://github.com/MITgcm/MITgcm/issues/735)
- It was decided to pass a string of command-line flags to genmake2 to be passed on to Tapenade, as this is more general, alllowing other flags if needed, but this can be changed if needed.
- the above set of subroutines represents the **minimum needed** to compile pkg/streamice with the `-defaultnocheckpoint` flag. This PR strictly focuses on streamice -- If the same flag is to be applied to the adjoint of other packages or other verification experiments, it is likely other subroutines will require development, and can be done in a separate PR.
- as of now in genmake2 the flag is passed to tapenade even in the Tangent Linear creation for completion, but this is not vital to keep.
- this PR does not include a change to `testreport` to pass flags to tapenade verification checks, but this can be provided if needed.

## Suggested addition to `tag-index`
o tools/genmake2
  - modify genmake2 in order to allow additional flags to be passed to tapenade with the argument "tap_extra" whose behaviour follows "taf_extra"
o tools/adjoint_options/adjoint_tap
  - add "default" tapenade flags here, rather than genmake2
o verification/halfpipe_streamice/build/genmake_local 
  - add tapenade flag 'defaultnocheckpoint'. This flag leads to shorter runtime of the adjoint model at the cost of larger memory.
o pkg/tapenade, eesupp/src/exch_tap_b.F
  - add _FWD and _BWD versions of manually written subroutine derivatives for: GLOBAL_SUM_TILE_RL, EXCH1_RL, EXCH1_RS, EXCH1_RL_CUBE, ACTIVE_READ_XY, ACTIVE_WRITE_XY

dngoldberg committed:
1st commit, adding Tapenade command line arguments through genmake2, adding fwd/bwd template functions to support -defaultnocheckpoint

dngoldberg committed:
replacing old file with new

dngoldberg committed:
remove _BWD and _FWD templates notused by streamice

dngoldberg committed:
Merge branch 'master' into branch_tap_cmdline

dngoldberg committed:
removing streamice includes which are not necessary for this PR

mjlosch committed:
fix use of macro and remove unused variables

mjlosch reviewed (commented):
This is a little out of my depth, but I have a couple of comments:
- I think it would be nice (but not absolutely necessary) to allow to pass a flag to tapenade also via testreport
- as an alternative to passing arguments to testreport, I tried to pass the new flag `TAPARG` via `genmake_local`, which overwrites whatever is passed on the command line, so we'd need an extra flag for that (eg. TAP_FLAGS or TAPENADE_FLAGS)
- does the order of TAPENADE-flags matter? If not (and it does not appear so), we could simplify the commands a little by combining all flags into one macro, e.g. `TAFARG = ${TAPARG} ${TAP_FLAGS} -ext "${TAPTOOLS}/flow_tap" -head "the_main_loop(fc)/(xx_genarr3d_dummy, xx_genarr2d_dummy, xx_gentim2d_dummy)"`, and then have more compact command lines: `\$(TAPENADECMD) -b \$(TAPARG) \$(AD_FILES) f95_test_mods.f90` (or `-d`); in case the arguments change, there's only one place to change this, and it's easier to read (?)
- the option `-defaultnocheckpoint` leads to a blow up in `halfpipe_streamice` on my platform (but that's maybe not directly relevant to this PR
- currently the flag `-defaultnocheckpoint` only works for `halfpipe_streamice`. Will there be appropriate error messages, when one tries with other experiments? Will it be hard to figure out, why an experiment fails that does not have the proper handwritten code?
- I see these warning messages:
```
(TC32) Conflicting numbers of arguments for procedure mds_read_sec_xz, expected 9, is here 11
(TC32) Conflicting numbers of arguments for procedure mds_read_sec_xz, expected 9, is here 11
(TC32) Conflicting numbers of arguments for procedure mds_read_sec_yz, expected 9, is here 11
(TC32) Conflicting numbers of arguments for procedure mds_read_sec_yz, expected 9, is here 11
```
Is this anything to worry about?

dngoldberg commented:
thanks @jm-c  and @mjlosch i assumed it would be needed for testreport but i don't know the code so might need guidance. for now i will look at martin's other comments

dngoldberg commented on eesupp/src/exch_tap_b.F:
```diff
@@ -121,6 +121,100 @@ SUBROUTINE EXCH1_RL_CUBE_B(
+c [llh] the following PUSHREAL8ARRAY could certainly be refined!
+c [ssg] for halfpipe_streamice
+c      CALL PUSHREAL8ARRAY(array, 26**2*2)
+c [ssg] for tutorial_global_oce_biogeo
```
these are left by laurent and @Shreyas911. I vote we remove them unless Shreyas does not want this.

dngoldberg committed:
Merge branch 'master' into branch_tap_cmdline

mjlosch committed:
add TAPENADE_FLAGS to genmake2 and offload long list of def-flags

- offload default flags to tools/adjoint_options/adjoint_tap
- add TAPENADE_FLAGS='-defaultnocheckpoint' to
  halfpipe_streamice/build/genmake_local for demonstration purposes

dngoldberg committed:
update flow_tap definitions to prevent tapenade warnings

dngoldberg commented:
hi @mjlosch and @jm-c just checking you are not waiting on me for anything in this PR? asking as i have just submitted yet another (though PR 961 is more of a bug fix.)

jm-c reviewed (changes requested):
Changes look OK, but few things to improve:

There are few more subroutines in several files, but these additional S/R are not listed at the top of the file, so that now this list (at the top) is incomplete. This is the case for `exch_tap_b.F` and the 2 active read/write files. And regarding `stubs_tap_adj.F`, I think it would be good to list, at the top, all the routines that this file contains, so that it's easier to quickly know what is there.

Regarding genmake2 and Makefile:

1. In the `Makefile`, the arguments that are passed to Tapenade are in `TAPARG`, but TAPARG does not appear in the AD-optfile `adjoint_tap`. So the only way to know how TAPARG is set is to search for in `genmake2` (> 4000 lines). By contrast, for TAF, the 3 variables that contain the TAF arguments (i.e., AD_/FTL_/SVD_TAF_FLAGS) are all set from the AD-optfile `adjoint_default`, so I don't need to dig into `genmake2`.
2. the new variables (TAPARG and/or TAPENADE_FLAGS) could be added to the lists "gm_s#" in `genmake2` to be printed to "genmake_state".
3. the new command line option "-taparg" description is not very precise, since it does not replace the "usual" tapenade arguments ( -ext ... -head ...) but add extra argument(s). There is also a typo where "-taparg" is processed (see blow).

A quick way to deal with these `genmake2` issues would be to remove the command-line option and, in the Makefile, rename `TATARG` to `TAPENADE_FLAGS`. An other alternative would be to mimic what the "-taf_extra" option is doing (except this one is not listed in the "-help").

jm-c commented:
@dngoldberg If you want to keep the command-line option, I think doing it the same/similar way as "-taf_extra" (e.g., "-tap_extra" ?) would be perfectly fine. We could think of other/alternative refinements, but these could come after and be discussed later.

dngoldberg committed:
Merge branch 'master' into branch_tap_cmdline

dngoldberg committed:
adding contents list (or adding to contents) for all new "manual derivative" SRs

dngoldberg committed:
replace "taparg" argument by "tap_extra" (with documentation). Add tap_extra to genmake2 "state". have "tap_extra" and "tapenade_flags" separately in tapenade call, rather than combining into a single variable.

dngoldberg commented on tools/genmake2:
```diff
@@ -3750,11 +3752,11 @@ if test $CAT_SRC_FOR_TAF = 1 ; then
 cat >>$MAKEFILE <<EOF
 adj_tap_all.$FS: \$(AD_FILES) f95_test_mods.f90
-	\$(TAPENADECMD) -b \$(TAPARG) \$(AD_FILES) f95_test_mods.f90
+	\$(TAPENADECMD) -b \$(TAPENADE_FLAGS) \$(TAP_EXTRA) \$(AD_FILES) f95_test_mods.f90
```
@mjlosch it compiled OK, but i will make ths change.

dngoldberg committed:
Reorder TAPENADE command flags in genmake2

dngoldberg commented:
@jm-c i have added contents tables in the relevant source files, and modified genmake2 to have a command line argument tap_extra, and i have done my best to make the functionality perfectly equivalent to taf_extra. (should tapenade_flags be part of the genmake2 state as well? it is not currently.)

on github_actions, there is now a breaking change, since the halfpipe_streamice example uses defaultnocheckpoint. This is likely an order of operations change. The accuracy is now 10 digits, so it is failing. Im not sure on what machine the new file should be produced.

jm-c commented:
@dngoldberg regarding changes to reference output file, I can take care of this (generate a new one on reference platform, i.e., on baudelaire for Tapenade tests).

dngoldberg committed:
Update genmake2 state to include TAPENADECMD and TAPENADE_FLAGS

jm-c committed:
Merge branch 'master' into branch_tap_cmdline

jm-c committed:
top list of S/R: fix ordering

jm-c committed:
minor edits

more consistent upper/lower case

jm-c committed:
new reference output (with -defaultnocheckpoint)

jm-c committed:
adjust the "help" info

jm-c reviewed (approved):
This looks good. I made some suggestions about the updated Tapenade ad-optfiles.

mjlosch committed:
rename local variable more consistently, change order of flags

jm-c committed:
fix spelling

replace "DIFF_FLAGSS" with simply "DIFF_FLAGS"

jm-c committed:
remove double report of TAPENADECMD

In genmake_state, print only once the value of TAPENADECMD

mjlosch reviewed (approved):
LGTM

jm-c committed:
Merge branch 'master' into branch_tap_cmdline

jm-c committed:
document more supported options for Tapenade
